### ISSUE TITLE
container_oom_events_total always returns 0

### ISSUE THREAD

**Issue by darkl0rd** [google/cadvisor#3015]
Running Docker (swarm), when OOM events occur the counter never increases. For reference, the node-exporter metric (node_vmstat_oom_kill) does increase.

Running cAdvisor v0.43.0.

**Comment by darkl0rd** [comment]
Just tested this with 0.44.0 - issue persists.

**Comment by tasiotas** [comment]
I have the same issue.

One of my container is running out of memory, I can see OOM event in syslog and kmsg, but `container_oom_events_total` is always 0. Any clues how to get it working, or other way to detect OOM in containers?

compose.yml
```
  cadvisor:
    image: gcr.io/cadvisor/cadvisor:v0.46.0
    container_name: cadvisor
    network_mode: host
    user: root
    privileged: true
    healthcheck:
      disable: true
    restart: unless-stopped
    ports:
      - '8080:8080'
    volumes:
      - /:/rootfs:ro
      - /var/run:/var/run:rw
      - /sys:/sys:ro
      - /var/lib/docker/:/var/lib/docker:ro
    command:
      - --url_base_prefix=/bRs47jH13fdsBFMQ93/cadvisor
      - --housekeeping_interval=15s
      - --docker_only=true
      - --store_container_labels=false
      - --enable_metrics=disk,diskIO,cpu,cpuLoad,process,memory,network,oom_event
    devices:
      - /dev/kmsg:/dev/kmsg
```

**Comment by Stolr** [comment]
Having the same issue here.

Event  not showing up in kubernetes also.
```
 Last State:     Terminated
      Reason:       Error
      Exit Code:    137
```

Any idea how to solve this ? 

Thanks !

**Comment by gfonseca-tc** [comment]
Also having the same issue with cAdvisor 0.46.0

**Comment by chengjoey** [comment]
Has there been any progress on this bug? I am encountering the same issue on k8s.

![image](https://user-images.githubusercontent.com/30427474/226533204-60effa2d-1142-4953-ba76-fa14304360b3.png)

describe go-demo pod:
![image](https://user-images.githubusercontent.com/30427474/226533339-cd7c38f4-abae-410e-8154-83ea0f76b67c.png)

**Comment by chengjoey** [comment]
In k8s, new containers are always created to replace oomkilled containers. Therefore, the `container_oom_events_total` metric will always be 0. I tried to keep the deleted containers due to oomkilled and was able to query the metrics.

![image](https://user-images.githubusercontent.com/30427474/226938105-7a6e4288-1cfc-4bf8-824d-462a44286a43.png)

In addition, if the cluster is running on minikube+docker, the containername obtained from `/dev/kmsg` will have an additional prefix of `/docker/{{id}}`, which does not match the containername watched. Therefore, the metrics will always be 0.
![image](https://user-images.githubusercontent.com/30427474/226938818-2924eed5-f307-4f60-9914-dc1705930cce.png)
![image](https://user-images.githubusercontent.com/30427474/226939067-57671f2f-de55-4c3a-8691-4034cfb1f7c2.png)

**Comment by NesManrique** [comment]
Hitting this issue in kubernetes as well. Commenting for visibility.

**Comment by tsipo** [comment]
I have done various tests of OOMKills under Kubernetes. I have - so far - seen only one use-case where I have observed `container_oom_events_total > 0` (specifically `container_oom_events_total == 1`).
An OOMKill does not mean in all cases deletion of the container. The main process of the container (aka `pid 1`) may fork one or more Linux processes (actually fork, not using the `exec` command). If one of these other processes gets OOMKilled and this does not cause `pid 1` to exit as well (at least not until the next cAdvisor scrape), the container will continue to live and you'll see `container_oom_events_total == 1`.

**Comment by pschichtel** [comment]
@tsipo track the referenced PR #3278 or help pushing it over the line.

**Comment by tsipo** [comment]
> @tsipo track the referenced PR #3278 or help pushing it over the line.

See https://github.com/google/cadvisor/pull/3278#issuecomment-2083587616

**Comment by rd-yan-farba** [comment]
Till now the problem persists.

**Comment by frittentheke** [comment]
@Creatone @bobbypage could I gently drag you to this very issue here about issue with OOM metrics?
Since there is a PR potentially fixing this issue, please see the recent comments there: https://github.com/google/cadvisor/pull/3278#issuecomment-2292982675


Is there anything to be done to get this problem addressed / the PR reviewed?

**Comment by tsipo** [comment]
> I have done various tests of OOMKills under Kubernetes. I have - so far - seen only one use-case where I have observed `container_oom_events_total > 0` (specifically `container_oom_events_total == 1`). An OOMKill does not mean in all cases deletion of the container (which will deregister its `container_oom_events_total`). The main process of the container (aka `pid 1`) may fork one or more Linux processes (actually fork, not using the `exec` command). If one of these other processes gets OOMKilled and this does not cause `pid 1` to exit as well (at least not until the next cAdvisor scrape), the container will continue to live and you'll see `container_oom_events_total == 1`.

An update on my previous comment: k8s `1.28` has enabled cgroup grouping (assuming `cgroups v2`) - see [here](https://github.com/kubernetes/kubernetes/pull/117793). That means the the use-case I have mentioned of "hidden OOMKills" (OOMKill of a container process which is not `pid 1`) should not happen anymore for k8s >= `1.28` and `cgroups v2`.

**Comment by chengjoey** [comment]
> An update on my previous comment: k8s `1.28` has enabled cgroup grouping (assuming `cgroups v2`) - see [here](https://github.com/kubernetes/kubernetes/pull/117793). That means the the use-case I have mentioned of "hidden OOMKills" (OOMKill of a container process which is not `pid 1`) should not happen anymore for k8s >= `1.28` and `cgroups v2`.

regardless of the k8s version or whether `cgroups v2` is used, as long as OOM causes the container to be rebuilt, the OOM metric is lost. I think kubelet or container-manager will still monitor the system OOM events, then kill the container and rebuild it, then the OOM metric will be lost

**Comment by SebastienTolron** [comment]
On my side, upgrade conmon and solved the issue ( minimum debian 12)

**Comment by sellers** [comment]
Essentially, in a DevSecOps world, engineers want to be able to track container applications behavior when their memory usage exceeds the memory requests and kubelet has to step in to say, "whoah".  That turns into an OOM (137 as stated above).  
If that happens over X time during a period of Y, as an engineer, I want to know so that I can address performance issues or business SLA/continuity issues.

e.g. I want to have a query like
```sum(increase((container_oom_events_total{namespace!="playground"}[5m]))>1) by (container_label_X, image_id)```

or something similar

missing-container-metrics offers this exact metric, albeit 3 years old and littered with CVEs now.  

to that, I am unsure what container_oom_events_total is intended for if not the above user story. Please do help all of us who are following this thread if we are misunderstanding the intention of this metric.

**Comment by frittentheke** [comment]
Using OOM event to properly set the resource requirements and limits is even one of the key things Google mentions in their famous Autopilot paper: https://research.google/pubs/autopilot-workload-autoscaling-at-google-scale/

TL;DR: OOM events are a very important signal at scale.

**Comment by jywhite1** [comment]
I'm struggling with the same issue here, following

**Comment by ringerc** [comment]
To work around this issue with OOM events and cgroups v2, I'm using `kube-state-metrics` as advised above, with queries like

```promql
# Detect recent OOMs
((time() - kube_pod_container_status_last_terminated_timestamp) < 30)
and on (uid,container)
(kube_pod_container_status_last_terminated_exitcode == 137)
```

and

```promql
# OOMs counted over a lookback window
count_over_time(
    (
        ((time() - kube_pod_container_status_last_terminated_timestamp) < 30)
        and on (uid,container)
        (kube_pod_container_status_last_terminated_exitcode == 137)
    )[1h:30s]
)
```

**Comment by frittentheke** [comment]
@dims would you kindly elaborate why this is not planned?

@ringerc solution is unfortunately a hacky and racy workaround (very nice of him to share is though) that does NOT solve the underlying bug.

**Comment by dims** [comment]
@frittentheke there isn't anyone to do the work

**Comment by frittentheke** [comment]
> @frittentheke there isn't anyone to do the work


Does not make the issue invalid though.

**Comment by dims** [comment]
Excellent point.

**Comment by frittentheke** [comment]
> Excellent point.

I am also wondering about the open PR https://github.com/google/cadvisor/pull/3278 ... apparently someone did the work (even though it needs a rebase)

**Comment by ringerc** [comment]
**TL;DR: this metric seems inherently broken and should probably be replaced with one that omits the `id` and `name` dimensions, counting OOMs dimensioned by pod and by container name.**

The concern with the above PR appears to be the need to keep old container state around for some period of time and report it after the container terminates. This is not done for other container state; cadvisor "forgets" the container state as soon as the container terminates. There are concerns about the impact this will have on cadvisor's memory use when there's high container churn.

I internally `containerd` is allocating a new container-id for each re-execution, and that's what cadvisor uses for the label. You won't see any container at all in `crictl ps` when the pod's container is in crash loop back-off. This can be observed by getting a shell on the kube node and running something like `while : ; do crictl ps |grep -i oom; sleep 0.1; done`  . Each container iteration gets a new container-id at containerd level.

So the metric `container_oom_events_total` doesn't really make sense with an `id` dimension, given that containerd will re-create it each time.

So it seems to me that **this metric probably needs to be deprecated and replaced with a new one that omits the `id` and `name` labels, instead being tracked at `pod` level with a `container` dimension for the container name**. 

For testing try these pod specs:

```
# leader process OOMs
kubectl apply --server-side -f - <<'__END__'
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: oomtest
  name: oomtest
spec:
  replicas: 1
  selector:
    matchLabels:
      app: oomtest
  template:
    metadata:
      labels:
        app: oomtest
    spec:
      containers:
      - command:
        - sh
        - -c
        - 'sleep 30s; echo "starting"; x="padding"; while : ; do x="${x}${x}"; sleep 0.1; done'
        image: busybox:latest
        name: oomtest
        resources:
          limits:
            memory: 5Mi
          requests:
            memory: 5Mi
      restartPolicy: Always
__END__
```

and this one that _should_ OOM a child process:

```
# Child process OOMs, leader does not
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: oomtestfork
  name: oomtestfork
spec:
  replicas: 1
  selector:
    matchLabels:
      app: oomtestfork
  template:
    metadata:
      labels:
        app: oomtestfork
    spec:
      containers:
      - command:
        - sh
        - -c
        - |
          sleep 30s; echo "starting"; while : ; do sh -c 'x="padding"; while : ; do x="${x}${x}"; sleep 0.1; done'; echo oomd; sleep 0.1; done
        image: busybox:latest
        name: oomtest
        resources:
          limits:
            memory: 50Mi
          requests:
            memory: 50Mi
      restartPolicy: Always
```

A PromQL query like

```
last_over_time(container_oom_events_total{namespace="default",container="oomtest"}[30m])
```

will return data like

```
container_oom_events_total{container="oomtest", endpoint="https-metrics", id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-poddd9e8754_f83a_4eab_b242_cee7138411b2.slice/cri-containerd-8b04ae69de5c48643d418ff82da46f350d483e234c011af969adb88af7bd8b4b.scope", image="docker.io/library/busybox:latest", instance="172.18.0.2:10250", job="kubelet", metrics_path="/metrics/cadvisor", name="8b04ae69de5c48643d418ff82da46f350d483e234c011af969adb88af7bd8b4b", namespace="default", node="edbpgai-worker3", pod="oomtest-86889bb6d5-tnrqh", service="kubelet"}
```

these vary by `id` and `name` with each execution as seen with `max by (id, name) (last_over_time(container_oom_events_total{namespace="default",container="oomtest"}[30m]))`

```
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod58e4e54d_9d83_4569_8bcd_aad3632440d4.slice/cri-containerd-9bbbcedbef212c911902c3bdf6eb9f78ae687cc7305200b404608a325231ddcc.scope", name="9bbbcedbef212c911902c3bdf6eb9f78ae687cc7305200b404608a325231ddcc"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod8c70014e_e67c_4f45_8c53_1686af542ac2.slice/cri-containerd-45c14ed2f23132b282339a4b1525c8b70bc739a03d9347eafa86b4d81edfe94f.scope", name="45c14ed2f23132b282339a4b1525c8b70bc739a03d9347eafa86b4d81edfe94f"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod58e4e54d_9d83_4569_8bcd_aad3632440d4.slice/cri-containerd-b9795d7fc7a6a35b557121695a2659c1acc2c10f20d8609e904b95d38b4ed921.scope", name="b9795d7fc7a6a35b557121695a2659c1acc2c10f20d8609e904b95d38b4ed921"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod8c70014e_e67c_4f45_8c53_1686af542ac2.slice/cri-containerd-35c000a5909e17117cdedfca07dd4309ed14dd398ccba7f0a3632c41f7d81aeb.scope", name="35c000a5909e17117cdedfca07dd4309ed14dd398ccba7f0a3632c41f7d81aeb"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod58e4e54d_9d83_4569_8bcd_aad3632440d4.slice/cri-containerd-394cf7ba19ba650365171e24465e01ee19c488a8dfba6de66c43d8531a2eaa6f.scope", name="394cf7ba19ba650365171e24465e01ee19c488a8dfba6de66c43d8531a2eaa6f"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod8c70014e_e67c_4f45_8c53_1686af542ac2.slice/cri-containerd-2ed8741f31f94342c16e4bd083c84e4cc6b266a6d8a77b1a90fc1f1797ccbfed.scope", name="2ed8741f31f94342c16e4bd083c84e4cc6b266a6d8a77b1a90fc1f1797ccbfed"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod58e4e54d_9d83_4569_8bcd_aad3632440d4.slice/cri-containerd-ed83a69e0b135ecdd34a571f7f1f5c8d744468a841abf1ec7a03e53437a3c14f.scope", name="ed83a69e0b135ecdd34a571f7f1f5c8d744468a841abf1ec7a03e53437a3c14f"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod8c70014e_e67c_4f45_8c53_1686af542ac2.slice/cri-containerd-dee90a0319ca798fa34a8b542c03f2d4c2bbe228b3cf84ae723127511446be9b.scope", name="dee90a0319ca798fa34a8b542c03f2d4c2bbe228b3cf84ae723127511446be9b"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod58e4e54d_9d83_4569_8bcd_aad3632440d4.slice/cri-containerd-94dbc99d3227d7251ae4fe4bb48ccdddb423d0db393a16ef17c18c08a78d11a4.scope", name="94dbc99d3227d7251ae4fe4bb48ccdddb423d0db393a16ef17c18c08a78d11a4"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod8c70014e_e67c_4f45_8c53_1686af542ac2.slice/cri-containerd-a2fd30bbefb0c1c91094cb8f42a26112e865f20748f278188ff134f1699a9185.scope", name="a2fd30bbefb0c1c91094cb8f42a26112e865f20748f278188ff134f1699a9185"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod58e4e54d_9d83_4569_8bcd_aad3632440d4.slice/cri-containerd-f9b9abe8582c08ca823b147f0dc959c3e79d56b060c93d07592303b409fa2d38.scope", name="f9b9abe8582c08ca823b147f0dc959c3e79d56b060c93d07592303b409fa2d38"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod8c70014e_e67c_4f45_8c53_1686af542ac2.slice/cri-containerd-263dec911e4703a4ebe7f9174903302881da7a52e56f589aa4c22850d5a1e931.scope", name="263dec911e4703a4ebe7f9174903302881da7a52e56f589aa4c22850d5a1e931"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod58e4e54d_9d83_4569_8bcd_aad3632440d4.slice/cri-containerd-014ed126e5972eaaf922b29b8b0907ddae0fa79bc643c5b9a8e0c34f5363ba08.scope", name="014ed126e5972eaaf922b29b8b0907ddae0fa79bc643c5b9a8e0c34f5363ba08"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod8c70014e_e67c_4f45_8c53_1686af542ac2.slice/cri-containerd-afcebab5e0a60e0f6b297ac55e452558b19f848caf8782769dc4658a52959f23.scope", name="afcebab5e0a60e0f6b297ac55e452558b19f848caf8782769dc4658a52959f23"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod58e4e54d_9d83_4569_8bcd_aad3632440d4.slice/cri-containerd-99bb09af399ae75d3cc4dbd6bf3a3442dea769b04bb3cbb8e02383f66fbcb145.scope", name="99bb09af399ae75d3cc4dbd6bf3a3442dea769b04bb3cbb8e02383f66fbcb145"}	0
{id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-pod8c70014e_e67c_4f45_8c53_1686af542ac2.slice/cri-containerd-a0be5ff8ab304cd57cfabada1053c6fc0ca38595a6340a1ae23ffe1f2261a5dd.scope", name="a0be5ff8ab304cd57cfabada1053c6fc0ca38595a6340a1ae23ffe1f2261a5dd"}	0
```

so **the whole concept of an oom count dimensioned by the container-id seems to be fairly broken** - unless there's a case where the container supervisor can recover from an OOM of one of its child processes and continue execution.

If I deploy this test workload https://github.com/densify-dev/oomkill/blob/main/examples/pod-oom-forked.yaml I see similar results for the metric

```
container_oom_events_total{container="memeater", endpoint="https-metrics", id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-podc6aba0fa_aecd_4419_9b53_adfae8f15b35.slice/cri-containerd-c1a54f2bf06c9e2e169ecf45f268be23e4db34ee787568c92ab40b90f44f3ebc.scope", image="docker.io/densify/memeater:latest", instance="172.18.0.2:10250", job="kubelet", metrics_path="/metrics/cadvisor", name="c1a54f2bf06c9e2e169ecf45f268be23e4db34ee787568c92ab40b90f44f3ebc", namespace="default", node="edbpgai-worker3", pod="main-oom-forked", service="kubelet"}	0
container_oom_events_total{container="memeater", endpoint="https-metrics", id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-podc6aba0fa_aecd_4419_9b53_adfae8f15b35.slice/cri-containerd-620e031af3bf1d89f7265fca38fe1e37c9d8203005dca2b996bb5dd1cdb5f02f.scope", image="docker.io/densify/memeater:latest", instance="172.18.0.2:10250", job="kubelet", metrics_path="/metrics/cadvisor", name="620e031af3bf1d89f7265fca38fe1e37c9d8203005dca2b996bb5dd1cdb5f02f", namespace="default", node="edbpgai-worker3", pod="main-oom-forked", service="kubelet"}	0
container_oom_events_total{container="memeater", endpoint="https-metrics", id="/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-podc6aba0fa_aecd_4419_9b53_adfae8f15b35.slice/cri-containerd-f634656f640f87e68f76137c73126b6e6a76c6cd3fb53ea86f435ed2d0028412.scope", image="docker.io/densify/memeater:latest", instance="172.18.0.2:10250", job="kubelet", metrics_path="/metrics/cadvisor", name="f634656f640f87e68f76137c73126b6e6a76c6cd3fb53ea86f435ed2d0028412", namespace="default", node="edbpgai-worker3", pod="main-oom-forked", service="kubelet"}
```

... it's still not incrementing it.

I added some more notes on the PR https://github.com/google/cadvisor/pull/3278#issuecomment-3698141231 on a possible approach using the containerd event stream.

**Comment by ringerc** [comment]
It seems like using the [CRI event stream](https://github.com/google/cadvisor/pull/3278#issuecomment-3698141231) would make sense unless I'm missing something blindingly obvious. The events already have the information required, including labels that map events directly back to the pod and container name so there should be no new state tracking required in cadvisor.

But `cadvisor` doesn't seem to use the [cri-api](https://github.com/kubernetes/cri-api), and that's surprising to the point where I feel like I must be missing something. Even if I `go mod vendor` in `master` and `grep` for `cri-api` there's nothing. I had hoped it'd be (relatively) practical to just add support for the cri-api `pkg/apis/runtime/v1/api.proto` 's [`GetEventsRequest`](https://github.com/kubernetes/cri-api/blob/e9948afd9e4958f45b299d2af6fca690577fb7ab/pkg/apis/runtime/v1/api.proto#L2006) to stream and filter container events, watching for OOMs. But it seems containerd doesn't use the CRI API at all.

Is there a known reason it doesn't use the CRI API? Instead it looks like [there's an alpha feature for Kubelet to expose metrics directly based on CRI without using cadvisor](https://kubernetes.io/docs/reference/instrumentation/cri-pod-container-metrics/) but it doesn't seem to have been stabilised despite being introduced in kube 1.23. It's not clear from https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md what the progress is, but it doesn't seem like the CRI-based metrics will imminently replace cadvisor; the last meaningful activity was for kube 1.29. Seems there's [some blockage on containerd support](https://github.com/kubernetes/kubernetes/issues/133884) though [with recent progress](https://github.com/containerd/containerd/pull/10691) so maybe it'll go somewhere soon.

**If there's no compelling reason not to, it'd IMO make sense for `cadvisor` to consume the CRI API for this data, and maintain counters in a race-free manner that way. The metric should probably also be deprecated in favour of one that omits `name` and `id` labels.**

**Comment by ringerc** [comment]
See also https://github.com/google/cadvisor/pull/3278#issuecomment-3698141231

**Comment by ringerc** [comment]
I looked at this a bit more closely.

The main findings:

For using cri-api:

* containerd currently doesn't connect to kubernetes cri-api at all; it uses specific container runtime apis. So a whole cri-api client would be required for this.
* CRI `ContainerEventType` is `CREATED`/`STARTED`/`STOPPED`/`DELETED`. OOM can be inferred from a `STOPPED` event whose `ContainerStatus.Reason` == `"OOMKilled"`. This relies on the runtime correctly populating it from cgroup/exit data. **It will only see OOMs that terminate the container**. Non-leader child OOMs that TaskOOM would catch are not seen.
* cadvisor supports more than k8s - it covers Docker Swarm etc too. Using k8s cri-api wouldn't work for that.
* `k8s.io/cri-api` is a k8s staging module. cadvisor is vendored into kubernetes. So there are some potentially messy dependency cycles involved here.

For using existing CRI-specific APIs instead:

* Containerd's `TaskOOM` carries only `ContainerID` and the container entry will already be destroyed preventing lookup, so `containerd` would need to maintain its own state tracking container-id -> namespace & pod. 
* Would only work for containerd; other runtimes would need different implementations that may or may not be viable.

Given all this, I'm figuring this `cadvisor` metric is basically unfixably broken; it'd require an excessively complex fix with a bunch of extra per-runtime work to work around deficiencies in the data sources and cadvisor's own design. It should really just be removed.

Kubelet is slowly moving to expose pod/container stats from CRI directly, bypassing cadvisor (KEP-2371); I'm hoping this provides a better solution.
